util/net_help.c

```diff
@@ -1191,6 +1217,44 @@ void* outgoing_ssl_fd(void* sslctx, int fd)
 #endif
 }
 
+/** set the authname on an SSL structure, SSL* ssl */
+int set_auth_name_on_ssl(void* ssl, char* auth_name)
+{
+	if(!auth_name) return 1;
+#ifdef HAVE_SSL
+	(void)SSL_set_tlsext_host_name(ssl, auth_name);
+#else
+	(void)ssl;
+#endif
+#ifdef HAVE_SSL_SET1_HOST
+	SSL_set_verify(ssl, SSL_VERIFY_PEER, NULL);
+	/* setting the hostname makes openssl verify the
+	 * host name in the x509 certificate in the
+	 * SSL connection*/
+	if(!SSL_set1_host(ssl, auth_name)) {
+		log_err("SSL_set1_host failed");
+		return 0;
+	}
+#elif defined(HAVE_X509_VERIFY_PARAM_SET1_HOST)
+	/* openssl 1.0.2 has this function that can be used for
+	 * set1_host like verification */
+	if(auth_name) {
+		X509_VERIFY_PARAM* param = SSL_get0_param(ssl);
+#  ifdef X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS
+		X509_VERIFY_PARAM_set_hostflags(param, X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
+#  endif
+		if(!X509_VERIFY_PARAM_set1_host(param, auth_name, strlen(auth_name))) {
+			log_err("X509_VERIFY_PARAM_set1_host failed");
+			return 0;
+		}
+		SSL_set_verify(ssl, SSL_VERIFY_PEER, NULL);
+	}
+#else
+	verbose(VERB_ALGO, "the query has an auth_name, but libssl has no call to perform TLS authentication");
```

Comment:
Are you sure? This would break people's configuration for upstream to public resolvers if their openssl happens to be older?  I mean this routine is used there mostly, not so much for DNSTAP.

Reply:
Right. I didn't consider the non-dnstap use. That will indeed break current deployments, making it a bad idea.

Reply:
The current code does not change unbound's behaviour for dns over tls. But shares some code with dnstap. I guess we can close it here, unless you want to change the dnstap behaviour (only)?

Reply:
Fix in https://github.com/NLnetLabs/unbound/pull/164/commits/398e2601454ec26e91a2d5119739545b4eb66054 .